File: moderate/flavius_josephus.rb
Node = Struct.new(:value, :next)
def list_push(value, node)
  Node.new(value, node)
end

File.open(ARGV[0]).each_line do |line|
  n, m = line.split(',').map(&:to_i)

  tail = Node.new(n - 1, nil)
  list = tail
  (n - 2).step(0, -1) { |x| list = list_push(x, list) }
  tail.next = list
  list = tail

  s = []
  n.times do
    (m - 1).times { list = list.next }
    s << list.next.value
    list.next = list.next.next
  end
  puts s.join(' ')
end

File: moderate/flavius_josephus.lua
for line in io.lines(arg[1]) do
  n, m = line:match("(%d+),(%d+)")
  tail = {next = nil, value = n-1}
  list = tail
  for i = n-2, 0, -1 do list = {next = list, value = i} end
  tail.next = list
  list = tail
  for i = 1, n do
    for _ = 1, m-1 do list = list.next end
    if i > 1 then io.write(' ') end
    io.write(list.next.value)
    list.next = list.next.next
  end
  io.write('\n')
end
